### PR TITLE
Add Inventory Check Handler

### PR DESCRIPTION
**Pull request by ZaneYork**

### CODE REVIEW

File: CubeModLoader/DLL.h [hunk 1]
```diff
@@ -17,6 +17,7 @@ class DLL
         // Callbacks
         FARPROC HandleChat;
         FARPROC HandleP2PRequest;
+        FARPROC HandleInventoryCheck;
```
Comment (ChrisMiuchiz) [inline CubeModLoader/DLL.h]: This name isn't consistent with "CheckInventoryFullHandler". Is this supposed to be "HandleCheckInventoryFull"?

File: CubeModLoader/callbacks/CheckInventoryFullHandler.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+void* CheckInventoryFullHandler_ptr = (void*)&CheckInventoryFullHandler;
+
+void* ASMCheckInventoryFullHandler_jmpback;
+void* ASMCheckInventoryFullHandler_retn;
+void no_optimize ASMCheckInventoryFullHandler() {
+    asm(PUSH_ALL
+
+        PREPARE_STACK
+        "sub rsp, 0x0F \n"
```
Comment (ChrisMiuchiz) [inline CubeModLoader/callbacks/CheckInventoryFullHandler.h]: Why is it necessary to sub rsp, 0x0F? If there isn't a good reason for this, I suspect this will actually cause problems. For example, PREPARE_STACK aligns the stack to 16 bytes and allocates 0x28 on the stack (RESTORE_STACK uses the missing 8 bytes, so this is still aligned to 16 bytes). If the stack is no longer aligned to 16 bytes, instructions in mods such as "movaps" can cause an exception because they expect offsets to be aligned.

tl;dr: I don't think you should ever have to manually allocate stack space before calling a function thanks to PREPARE_STACK, and this particular allocation is likely to cause crashes due to not being a multiple of 16.

Comment (ZaneYork) [reply CubeModLoader/callbacks/CheckInventoryFullHandler.h]: PREPARE_STACK allocated 0x30 stack space.It is only big enough for one parameter.0x20 for basic use,0x08 for ebx,0x08 for eax.This handler needs to pass two pointer,so we needs another 0x08 space for it.Due to the alignment,it is 0x0F

Comment (ChrisMiuchiz) [reply CubeModLoader/callbacks/CheckInventoryFullHandler.h]: Allocating 0x20 is just because the x86-64 calling convention requires it, but I don't think it's actually required in this program in practice.

x86-64 is different from x86-32. On x86-32, you'd almost always have to push arguments, but on x86-64, parameters are stored in RCX, RDX, R8, and R9. Since these functions don't take more than 4 parameters, they will all be held in those registers. That is to say, the cube::Creature* is in RCX, and the cube::Item* is in RDX, and they do not need additional space allocated for them. Have you tested removing the 0xF allocation?

Comment (ZaneYork) [reply CubeModLoader/callbacks/CheckInventoryFullHandler.h]: Yes,I tested.It crashed when I remove this instruction, may be this is not the actual reason to crash?

Comment (ZaneYork) [reply CubeModLoader/callbacks/CheckInventoryFullHandler.h]: I'll test it again next day .Good night.

Comment (ChrisMiuchiz) [reply CubeModLoader/callbacks/CheckInventoryFullHandler.h]: Did you also remove "add rsp, 0x0F \n"? If you didn't also remove that, that will result in a crash. Removing both instructions ("sub rsp, 0x0F" and "add rsp, 0x0F") does not crash for me. 

I did have an issue with the callback registered from a mod, but it turned out to be my mistake combined with a bug in GCC. If you don't specify a return type and return from the function in the mod, GCC may not actually return from the function, and simply let it NOP sled into the next one. Make sure yours also returns correctly.

**Comment by ChrisMiuchiz** [comment]
Were you able to get it working without the 15 byte allocation on your computer?

**Comment by ZaneYork** [comment]
Yes, it  works, so I had removed these useless instructions
